clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java

```diff
@@ -2925,4 +2766,14 @@ private void assertFutureFailed(Future<RecordMetadata> future) throws Interrupte
         }
     }
 
+    private void runUntil(Supplier<Boolean> condition) {
```

Comment:
Would you consider using a `long timeout` instead and move this to TestUtils? Perhaps we need this somewhere else too.

Reply:
It's a bit difficult to restructure these test cases with a timeout. Most of the logic is basically "run the sender until X happens, then verify Y." With a timeout, we cannot control the stopping condition.